kvserver: emit `AddSSTable` events via rangefeeds

erikgrinaker opened the pull request:
**testutils: add sstutil package**

This patch adds a `testutils/sstutil` package with SST-related test
utilities.

Release note: None

**kvserver: disconnect rangefeeds on MVCC history mutations**

This patch adds a field `MVCCHistoryMutation` to `ReplicatedEvalResult`,
which disconnects any rangefeeds that overlap with the command span.
Callers are expected to ensure there are no rangefeeds over such spans,
but if they fail to do so it is better to error out rather than silently
pretend like nothing happened. The field is set for `AddSSTable`,
`ClearRange`, and `RevertRange` requests.

This error is exposed via the `OnInternalError` callback for
`rangefeed.Factory`-based rangefeed clients. However, if this callback
is not set, these clients will silently stop processing events. This is
unfortunate, but follows from the API design. Most callers do not appear
to set such a callback, and it is left for the owning teams to update
the usage of the library with appropriate error handling. Furthermore,
this error detection requires knowledge about the new
`MVCCHistoryMutationError` error type, so older nodes in mixed-version
clusters will simply treat this as a retryable error.

Touches #70434.

Release note: None

**kvserver: emit AddSSTable events via rangefeeds**

This patch emits `AddSSTable` events across rangefeeds when the
ingestion was done with `WriteAtRequestTimestamp` enabled (since this
ensures they respect the closed timestamp). This setting is introduced
with 22.1, and callers must check the `MVCCAddSStable` version gate
before using it, so by extension this event is only emitted once the
entire cluster runs 22.1.

Clients built via `rangefeed.Factory` have a new `WithOnSSTable` option
that can be used to register a callback for these events. If such a
callback is not set, the rangefeed will run a catchup scan that includes
the values written by the `AddSSTable` request.

The entire SST is emitted in binary form, regardless of how the
registration span overlaps it -- it is up to callers to prune the SST
contents as appropriate. Previous values are not included for keys
replaced by the SST.

Resolves #70434.

Release note: None

---

Initial draft. A few outstanding questions:

1. ~~Changefeeds will error on these SST events. We currently do not expect to ingest SSTs into online tables, and thus we do not expect changefeeds to see these events. Now that we have MVCC-compliant `AddSSTable` we _could_ begin to use `AddSSTable` into online tables, and thus the changefeeds would need to handle this (note here that the SST events do not contain previous value diffs), but I figure we can cross that bridge when we get there. Does this make sense?~~

2. ~~Internal consumers that subscribe via `rangefeed.Factory` must register an `OnSSTable` handler, otherwise they will not be notified about such events. I'm not sure if we ever expect to ingest SSTs into e.g. online system spans and such, but we may want to consider sending such subscribers an error if they haven't registered `OnSSTable` rather than silently ignoring them.~~

3. These events are likely to be much larger than regular key/value events, and thus there is a greater risk of excessive memory usage since the buffer size per replica is 4096 events. The SSTs are already in memory due to Raft application, but buffering them here can cause them to stick around and pile up. We should consider adding memory accounting here, and use a store-wide memory limit rather than a per-replica fixed-size queue. See: #73616.

erikgrinaker committed:
kvserver: emit `AddSSTable` events via rangefeeds

This patch emits `AddSSTable` events across rangefeeds when the
ingestion was done with `WriteAtRequestTimestamp` enabled (since this
ensures they respect the closed timestamp). This setting is introduced
with 22.1, and callers must check the `MVCCAddSStable` version gate
before using it, so by extension this event is only emitted once the
entire cluster runs 22.1.

Clients built via `rangefeed.Factory` have a new `WithOnSSTable` option
that can be used to register a callback for these events. If such a
callback is not set, the rangefeed will run a catchup scan that includes
the values written by the `AddSSTable` request.

The entire SST is emitted in binary form, regardless of how the
registration span overlaps it -- it is up to callers to prune the SST
contents as appropriate. Previous values are not included for keys
replaced by the SST.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73487)
<!-- Reviewable:end -->

miretskiy reviewed (commented):
> Reviewed 10 of 18 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @erikgrinaker)

___
*[-- commits, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MqExWrrB0QtBvza4Hsa):*
I don't see anywhere in this commit that checks for MVCCAddSSTable setting.
What's the intended use case?  Do callers (i.e. rangefeed library users) ought to check
for this setting before enabled OnSST event?  Just trying to clarify correct usage.
___
*[pkg/kv/kvclient/rangefeed/config.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MqErRLNEW9DROYFdQJn:-MqErRLNEW9DROYFdQJo:b1llcgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f85b78aa0350effa92d07f5f406b651c88394b83/pkg/kv/kvclient/rangefeed/config.go#L115)):*
> ```Go
> }
> 
> // OnSST is called when a SSTable is ingested.
> ```

I was going to suggest that perhaps we should drop this option and SST handling since sst handling is not "trivial"...
But maybe just expand the comment to indicate how subtle OnSST needs to be.  In particular, the fact that it may
include keys outside span boundaries.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 367 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MqErdr3CHfRLbnn2v1J:-MqErdr3CHfRLbnn2v1K:bx2oxrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f85b78aa0350effa92d07f5f406b651c88394b83/pkg/kv/kvclient/rangefeed/rangefeed.go#L367)):*
> ```Go
> 				// FIXME(erikgrinaker): Callers who do not set onSSTable will be oblivious
> 				// to these events -- we may want to send such callers an error and force
> 				// them to do a catchup scan.
> ```

I like the idea of returning an error to cause catchup scan very much.
Or we need to provide a default impl that iterates sst, while ignoring keys outside span boundaries.  But
forcing catchup scan sounds a lot better.  the caller decides , if they know what they are doing, to ask for SSTs; 
otherwise things continue to work.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @miretskiy)

___
*[-- commits, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MqExWrrB0QtBvza4Hsa):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I don't see anywhere in this commit that checks for MVCCAddSSTable setting.
What's the intended use case?  Do callers (i.e. rangefeed library users) ought to check
for this setting before enabled OnSST event?  Just trying to clarify correct usage.
</blockquote></details>

It is up to `AddSSTable` RPC callers to check `MVCCAddSSTable` before using `WriteAtRequestTimestamp`. The server-side logic needs to respect `WriteAtRequestTimestamp` regardless of the version gate in order to handle the non-atomic rollout of version gates across the cluster, as described here:

https://github.com/cockroachdb/cockroach/blob/d27f579797e53d234e112d48370780ccfa6674e8/pkg/clusterversion/cockroach_versions.go#L50-L84

Once that gate starts being activated, all cluster nodes (including rangefeed consumers) will already be running 22.1, so even though they haven't observed the `MVCCAddSSTable` flip they should be able to consume these events correctly.
___
*[pkg/kv/kvclient/rangefeed/config.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MqErRLNEW9DROYFdQJn:-MqF04R85xtTKuufyUVR:b-dqxwil) ([raw file](https://github.com/cockroachdb/cockroach/blob/f85b78aa0350effa92d07f5f406b651c88394b83/pkg/kv/kvclient/rangefeed/config.go#L115)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I was going to suggest that perhaps we should drop this option and SST handling since sst handling is not "trivial"...
But maybe just expand the comment to indicate how subtle OnSST needs to be.  In particular, the fact that it may
include keys outside span boundaries.
</blockquote></details>

We need to pass SSTs across rangefeeds in 22.1 to prepare for streaming cluster-to-cluster replication. Aborting these and forcing catchup-scans on every `AddSSTable` would likely be too expensive and cause the replication stream to lag significantly (consider that e.g. index builds will be firing off these requests continually for several hours). So I think it'll have to be up to callers to decide if and how they want to handle them.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 367 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MqErdr3CHfRLbnn2v1J:-MqF19Tr-k0l9ulyQggb:bjkppzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f85b78aa0350effa92d07f5f406b651c88394b83/pkg/kv/kvclient/rangefeed/rangefeed.go#L367)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I like the idea of returning an error to cause catchup scan very much.
Or we need to provide a default impl that iterates sst, while ignoring keys outside span boundaries.  But
forcing catchup scan sounds a lot better.  the caller decides , if they know what they are doing, to ask for SSTs; 
otherwise things continue to work.
</blockquote></details>

Yeah, it's probably better to error rather than silently ignore them -- I'll just have to audit all of the current internal callers (i.e. non-changefeed users) to make sure they'll actually run catchup-scans when they receive errors.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @erikgrinaker, and @miretskiy)

___
*[pkg/kv/kvclient/rangefeed/config.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MqErRLNEW9DROYFdQJn:-MqF4ITl4dvYCsVy9nPO:bukx9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/f85b78aa0350effa92d07f5f406b651c88394b83/pkg/kv/kvclient/rangefeed/config.go#L115)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

We need to pass SSTs across rangefeeds in 22.1 to prepare for streaming cluster-to-cluster replication. Aborting these and forcing catchup-scans on every `AddSSTable` would likely be too expensive and cause the replication stream to lag significantly (consider that e.g. index builds will be firing off these requests continually for several hours). So I think it'll have to be up to callers to decide if and how they want to handle them.
</blockquote></details>

well, I think the assumption here is that replication stream will use this library; it may very well do so, but that's not a given.
So, I just want to make sure that the existing users of this library behave correctly.  That's why I sort of wanted to expand the comment.  It's not a simple "kv" even like before.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @miretskiy)

___
*[pkg/kv/kvclient/rangefeed/config.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MqErRLNEW9DROYFdQJn:-MqF587B7m50KvAJIPK-:bvo0nsj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f85b78aa0350effa92d07f5f406b651c88394b83/pkg/kv/kvclient/rangefeed/config.go#L115)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

well, I think the assumption here is that replication stream will use this library; it may very well do so, but that's not a given.
So, I just want to make sure that the existing users of this library behave correctly.  That's why I sort of wanted to expand the comment.  It's not a simple "kv" even like before.
</blockquote></details>

Yep, totally. But I think that if rangefeeds are to include SSTs, then those events should be accessible via this library as well. But yeah, we should look into how to best integrate with existing callers (probably error by default, as discussed elsewhere), and I'll  make sure to expand the comment.


<!-- Sent from Reviewable.io -->

dt commented on pkg/kv/kvserver/batcheval/cmd_add_sstable.go:
```diff
@@ -248,8 +259,10 @@ func EvalAddSSTable(
+				Data:          sst,
+				CRC32:         util.CRC32(sst),
+				Span:          roachpb.Span{Key: start.Key, EndKey: end.Key},
+				EmitRangefeed: args.WriteAtRequestTimestamp,
```
I wonder if it'd make sense to persist this bool as "AtWriteTimestamp" or something like that, and then below instead of skipping non-mvcc AddSSTables, go ahead and send them to the feed until we get to `ev.SST != nil` and then there, if the SST was _not_ AtWriteTimestamp, fail the feed because we got an unexpected violation of the mvcc semantics that feed is relying on?

erikgrinaker replied:
Yeah, I think that makes sense -- describe the SST itself, and move the rangefeed policy elsewhere. I'll update the PR.

erikgrinaker replied:
I think it might make sense to call this `AtRequestTimestamp` and defer the rangefeed policy elsewhere, such that the data describes the SST rather than prescribing downstream actions. Even though it may be a bit unfortunate to bury this policy deep in the bowels of Raft state machine application.

However, given the size of these SSTs, I'd rather not ship all of them through the rangefeed machinery unnecessarily. Also, we only propagate the write timestamp into the rangefeed, with the assumption that all MVCC timestamps in the SST are equal to it:

https://github.com/cockroachdb/cockroach/blob/b3a84a243eeb2c98455a17e51b79b17e9ac97ef1/pkg/kv/kvserver/replica_rangefeed.go#L617-L619

It seems unnecessary to add a boolean parameter here, and then assert that the passed parameter is always `true`. Shouldn't it be the caller's responsibility to only call the method when appropriate?

erikgrinaker replied:
Then again, maybe we could move the policy into `Processor.ConsumeSSTable` such that it ignores `AtWriteTimestamp == false`

dt replied:
My initial suggestion would be to move it further beyond raft, all the way to `*RangeFeed.processEvents`, where it would be an _error_ if an sst event arrived that was not marked as `AtRequestTimestamp`. If we were just ignoring them at that point I'd be more concerned about the cost of shipping them all the way through just to discard them, but I was imaging that if we kick out an assertion failure error or something when the very first arrives that tears down the feed then that cost isn't a big deal since it quickly, and importantly loudly, comes to a stop?

erikgrinaker replied:
Ok, but I think I'd prefer to assert that on the server-side at least, e.g. inside the rangefeed processor at `Processor.publishSSTable`, rather than waiting for it to make its way out to the client and assert it there. In the latter case, we'd have to implement this in all clients (not just the `Factory`-constructed ones which aren't always used), along with the necessary plumbing.

erikgrinaker replied:
Should we maybe do this for all other non-MVCC operations such `ClearRange` and `RevertRange` too? I.e. signal to the rangefeed that we applied a non-MVCC operation, and disconnect it?

miretskiy replied:
That sounds good to me.

erikgrinaker replied:
Done, see separate commit which disconnects rangefeeds with a permanent `MVCCHistoryMutationError` for operations that mutate MVCC history -- specifically `AddSSTable` when not writing at the request timestamp, `ClearRange`, and `RevertRange`. This does *not* apply to MVCC-compliant `AddSSTable` requests that write at the request timestamp, which are emitted across rangefeeds.

Note that for `rangefeed.Factory`-based clients, this is surfaced via `OnInternalError` due to the current API design -- unfortunately, most callers do not set this callback, and will silently ignore the error even though the rangefeed stops. The owning teams should add appropriate error handling here.

erikgrinaker committed:
kvserver: disconnect rangefeeds on MVCC history mutations

This patch adds a field `MVCCHistoryMutation` to `ReplicatedEvalResult`,
which disconnects any rangefeeds that overlap with the command span.
Callers are expected to ensure there are no rangefeeds over such spans,
but if they fail to do so it is better to error out rather than silently
pretend like nothing happened. The field is set for `AddSSTable`,
`ClearRange`, and `RevertRange` requests.

This error is exposed via the `OnInternalError` callback for
`rangefeed.Factory`-based rangefeed clients. However, if this callback
is not set, these clients will silently stop processing events. This is
unfortunate, but follows from the API design. Most callers do not appear
to set such a callback, and it is left for the owning teams to update
the usage of the library with appropriate error handling. Furthermore,
this error detection requires knowledge about the new
`MVCCHistoryMutationError` error type, so older nodes in mixed-version
clusters will simply treat this as a retryable error.

Release note: None

erikgrinaker committed:
testutils: add `sstutil` package

This patch adds a `testutils/sstutil` package with SST-related test
utilities.

Release note: None

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @miretskiy)

___
*[pkg/kv/kvclient/rangefeed/config.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MqErRLNEW9DROYFdQJn:-MsBYrKR0s67CHR9HBg9:b-mhd9cp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f85b78aa0350effa92d07f5f406b651c88394b83/pkg/kv/kvclient/rangefeed/config.go#L115)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Yep, totally. But I think that if rangefeeds are to include SSTs, then those events should be accessible via this library as well. But yeah, we should look into how to best integrate with existing callers (probably error by default, as discussed elsewhere), and I'll  make sure to expand the comment.
</blockquote></details>

Expanded the comment with relevant details.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 367 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MqErdr3CHfRLbnn2v1J:-MsBZPk1EQJQv4hBDf1q:bk875up) ([raw file](https://github.com/cockroachdb/cockroach/blob/f85b78aa0350effa92d07f5f406b651c88394b83/pkg/kv/kvclient/rangefeed/rangefeed.go#L367)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Yeah, it's probably better to error rather than silently ignore them -- I'll just have to audit all of the current internal callers (i.e. non-changefeed users) to make sure they'll actually run catchup-scans when they receive errors.
</blockquote></details>

Done, we run catchup scans unless callers set `OnSSTable`. This only applies to MVCC-compliant `AddSSTable`, i.e. with `WriteAtRequestTimestamp` set. For legacy `AddSSTable`, we emit a permanent `MVCCHistoryMutationError` that this client library exposes via `OnInternalError` (which unfortunately most callers do not currently handle).


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
I think this should be ready for review now, barring any test failures.

miretskiy reviewed (commented):
This is very nice.  I think @dt and @ajwerner should also do a pass.

> Reviewed 3 of 23 files at r3, 9 of 20 files at r5, 1 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-:-MskECAqEGjiuPCLddH-:b-9geedp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @erikgrinaker, and @miretskiy)

___
*[pkg/kv/kvclient/rangefeed/config.go, line 149 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-Msk6wuE5t9jWZVoU7dA:-Msk6wuE5t9jWZVoU7dB:b-gnkqgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08ca066b7c1c43a77d1b2b450a6379e66b2b741/pkg/kv/kvclient/rangefeed/config.go#L149)):*
> ```Go
> // Also note that AddSSTable requests that do not set the
> // WriteAtRequestTimestamp flag, possibly writing below the closed timestamp,
> // will cause affected rangefeeds to be disconnected with MVCCHistoryMutationError
> ```

nit: perhaps add `a terminal` before MVCCHIstoryMutationError?
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 283 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-Msk7jI40OZhNMxmnQkp:-Msk7jI40OZhNMxmnQkq:bkepmtx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08ca066b7c1c43a77d1b2b450a6379e66b2b741/pkg/kv/kvclient/rangefeed/rangefeed.go#L283)):*
> ```Go
> 		start := timeutil.Now()
> 
> 		// Note that the below channel send will not block forever because
> ```

not even sure which channel this refers anymore.  I suspect it's talking about producer/consumer blocking.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 288 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-Msk80YbBfA02fPPab99:-Msk80YbBfA02fPPab9A:bpyzv2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08ca066b7c1c43a77d1b2b450a6379e66b2b741/pkg/kv/kvclient/rangefeed/rangefeed.go#L288)):*
> ```Go
> 		// returning. Because of this, only processEvents may use the cancel
> 		// function generated below, since the caller must also consume the errCh
> 		// error.
> ```

This logic could probably be simplified somewhat by simply creating errCh w/ buffer for 1 error.
Probably something to do for another time.
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 293 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-Msk9oKG8nQbjmn7Oj2C:-Msk9oKG8nQbjmn7Oj2D:b1hbika) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08ca066b7c1c43a77d1b2b450a6379e66b2b741/pkg/kv/kvclient/rangefeed/rangefeed.go#L293)):*
> ```Go
> 			errCh <- f.client.RangeFeed(ctx, f.spans, ts, f.withDiff, eventCh)
> 		}); err != nil {
> 			rfCancel()
> ```

This actually makes me a bit sad.  I have to jump through the code to verify that we're not leaking context here.
(usually, you always have to call `defer rfCancel()` immediately after WithCancel -- and actually, I think we we might fail to call rfCancel in some code paths.)
 And even if you can convince me that processEvents will always call rfCancel() eventually, I'd say that it's still brittle.  To be clear: it's not your code per-say, I think it's because of how this errCh communication was setup.

I wonder why we have to use stopper here at all.  We already run this go routine under stopper (with cancel on quiesce).  I wonder if another go routine could help us out w/ a more straightforward code.
Something like:
```
rangefeedTask := func(ctx context.Context) error {
   return f.client.RangeFeed(ctx, f.spans, ts, f.withDiff, eventCh)
}
processEventsTask := func(ctx2 context.Context) error {
   return f.processEvents(ctx, frontier, eventCh)
}

err := ctxgroup.GoAndWait(ctx, rangefeedTask, processEventsTask)
```

I don't think we'd need errCh at all in this case.  Just return the error from any of the go routines.  That cancels the other go routine, and you get the original error. 
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 365 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MskBYc42XqzyC30nDNb:-MskBYc42XqzyC30nDNc:b53e3ak) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08ca066b7c1c43a77d1b2b450a6379e66b2b741/pkg/kv/kvclient/rangefeed/rangefeed.go#L365)):*
> ```Go
> 					rfCancel()
> 					<-errCh
> 					return errors.New("received unexpected rangefeed SST event with no OnSSTable handler")
> ```

do you think errors.AssertionFailedf might be a better match?
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 623 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MskBhm_1k9bvPbvdgJV:-MskBhm_1k9bvPbvdgJW:b-4te2km) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08ca066b7c1c43a77d1b2b450a6379e66b2b741/pkg/kv/kvserver/replica_application_state_machine.go#L623)):*
> ```Go
> 	// expected to ensure that no rangefeeds are currently active across such
> 	// spans, but as a safeguard we disconnect the overlapping rangefeeds
> 	// with a non-retriable error anyway.
> ```

:heart: the comment.  Thank you.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 660 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MskBv1_0GpKuFdSWlYZ:-MskBv1_0GpKuFdSWlY_:b330hjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08ca066b7c1c43a77d1b2b450a6379e66b2b741/pkg/kv/kvserver/replica_application_state_machine.go#L660)):*
> ```Go
> 			b.r.writeStats.recordCount(float64(added), 0)
> 		}
> 		if res.AddSSTable.AtWriteTimestamp {
> ```

either here or at handleSSTRaftMuLocked, do we need a TODO comment to add memory accounting/limits?
___
*[pkg/kv/kvclient/rangefeed/rangefeed_external_test.go, line 504 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MskDrTE-jIE_MDmbJka:-MskDrTE-jIE_MDmbJkb:brktuej) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08ca066b7c1c43a77d1b2b450a6379e66b2b741/pkg/kv/kvclient/rangefeed/rangefeed_external_test.go#L504)):*
> ```Go
> 	select {
> 	case <-checkpointC:
> 	case <-time.After(time.Second):
> ```

isn't it a bit aggressive? under stress?


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @dt, and @miretskiy)

___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 293 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-Msk9oKG8nQbjmn7Oj2C:-MslTajd9PrYG4bSX7PI:b-7rps8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08ca066b7c1c43a77d1b2b450a6379e66b2b741/pkg/kv/kvclient/rangefeed/rangefeed.go#L293)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This actually makes me a bit sad.  I have to jump through the code to verify that we're not leaking context here.
(usually, you always have to call `defer rfCancel()` immediately after WithCancel -- and actually, I think we we might fail to call rfCancel in some code paths.)
 And even if you can convince me that processEvents will always call rfCancel() eventually, I'd say that it's still brittle.  To be clear: it's not your code per-say, I think it's because of how this errCh communication was setup.

I wonder why we have to use stopper here at all.  We already run this go routine under stopper (with cancel on quiesce).  I wonder if another go routine could help us out w/ a more straightforward code.
Something like:
```
rangefeedTask := func(ctx context.Context) error {
   return f.client.RangeFeed(ctx, f.spans, ts, f.withDiff, eventCh)
}
processEventsTask := func(ctx2 context.Context) error {
   return f.processEvents(ctx, frontier, eventCh)
}

err := ctxgroup.GoAndWait(ctx, rangefeedTask, processEventsTask)
```

I don't think we'd need errCh at all in this case.  Just return the error from any of the go routines.  That cancels the other go routine, and you get the original error. 

</blockquote></details>

Nice, I quite like this -- updated the code. I definitely share your distaste for the initial attempt, and you're right that it would leak a context.

However, I believe `RunAsyncTask` does a lot of other stuff too, such as setting up tracing spans. But maybe that doesn't matter as much here. Wdyt @andreimatei?
___
*[pkg/kv/kvclient/rangefeed/rangefeed.go, line 365 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MskBYc42XqzyC30nDNb:-MslH0wT5b9sy6gvkkrR:belan7p) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08ca066b7c1c43a77d1b2b450a6379e66b2b741/pkg/kv/kvclient/rangefeed/rangefeed.go#L365)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

do you think errors.AssertionFailedf might be a better match?
</blockquote></details>

Yeah, I agree.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 660 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MskBv1_0GpKuFdSWlYZ:-MslI6wf0TpdnndRjqzI:bfff55m) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08ca066b7c1c43a77d1b2b450a6379e66b2b741/pkg/kv/kvserver/replica_application_state_machine.go#L660)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

either here or at handleSSTRaftMuLocked, do we need a TODO comment to add memory accounting/limits?
</blockquote></details>

I think it's probably a bit overkill, but added a comment on `handleSSTableRaftMuLocked`. I'd like to implement this in the near future anyway.
___
*[pkg/kv/kvclient/rangefeed/rangefeed_external_test.go, line 504 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/73487#-MskDrTE-jIE_MDmbJka:-MslIn4O6Z0lk0KwnHiQ:bx3j9um) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08ca066b7c1c43a77d1b2b450a6379e66b2b741/pkg/kv/kvclient/rangefeed/rangefeed_external_test.go#L504)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

isn't it a bit aggressive? under stress?
</blockquote></details>

Possibly. Bumped them to 3s.


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/kv/kvclient/rangefeed/rangefeed.go:
```diff
@@ -280,19 +280,16 @@ func (f *RangeFeed) run(ctx context.Context, frontier *span.Frontier) {
+			return f.client.RangeFeed(ctx, f.spans, ts, f.withDiff, eventCh)
+		}
+		processEventsTask := func(ctx context.Context) error {
+			return f.processEvents(ctx, frontier, eventCh)
```
That's nice.  Whatever run async task did, this function (run) already runs under that -- so spans, ctx should already be setup.

miretskiy replied:
You can probably move those definition outside the loop.

erikgrinaker replied:
>Whatever run async task did, this function (run) already runs under that -- so spans, ctx should already be setup.

Not entirely -- `RunAsyncTask` will fork a new tracing span for the rangefeed goroutine. Without that, these two goroutines will be sharing the same tracing span.

> You can probably move those definition outside the loop.

The rangefeed needs to close over the `ts` variable which is scoped to the loop. I don't think it's safe to share `frontier` between the goroutines.

miretskiy replied:
Ahh.. I missed the ts arg -- you're right.  But the frontier is not shared -- we wait for the previous go routine to finish before we start next one... Anyway, it's probably clearer to keep things as they are.

erikgrinaker commented:
@dt @ajwerner Interested in doing a pass over this?

erikgrinaker commented:
TFTR! For posterity, @ajwerner and @dt deferred to @miretskiy.

bors r=miretskiy
